### PR TITLE
Use TestApplication

### PR DESCRIPTION
**Pull request by pakrym**
https://github.com/aspnet/Hosting/pull/1511

### CODE REVIEW

**Comment by jkotalik** [comment]
Something must have happened to the vsts webhook. Seems the tests passed there.

File: test/IISExpress.FunctionalTests/InProcess/StartupTests.cs [hunk 1]
```diff
@@ -163,16 +162,9 @@ public static TestMatrix TestVariants
         }
 
         // Defaults to inprocess specific deployment parameters
-        public static IISDeploymentParameters GetBaseDeploymentParameters(string site = "InProcessWebSite")
+        public IISDeploymentParameters GetBaseDeploymentParameters(TestApplication application = null)
```
Comment (jkotalik) [inline test/IISExpress.FunctionalTests/InProcess/StartupTests.cs]: Can we remove this function and use the fixture instead?

File: test/IISExpress.FunctionalTests/UpgradeFeatureDetectionTests.cs [hunk 2]
```diff
@@ -9,18 +9,21 @@
...
-        private readonly string _isWebsocketsSupported = Environment.OSVersion.Version >= new Version(6, 2) ? "Enabled" : "Disabled";
+        private readonly PublishedSitesFixture _fixture;
 
-        public UpgradeFeatureDetectionTests(ITestOutputHelper output) : base(output)
+        public UpgradeFeatureDetectionTests(PublishedSitesFixture fixture)
         {
+            _fixture = fixture;
         }
 
+        private readonly string _isWebsocketsSupported = Environment.OSVersion.Version >= new Version(6, 2) ? "Enabled" : "Disabled";
```
Comment (jkotalik) [inline test/IISExpress.FunctionalTests/UpgradeFeatureDetectionTests.cs]: nit move above constructor.

**Comment by Tratcher** [comment]
How much faster is it? What's the test run time before and after?

**Comment by pakrym** [comment]
Publishing via TestApplication is 0.1sec vs >1sec for dotnet publish.

File: test/Common.FunctionalTests/PublishedSitesFixture.cs [hunk 3]
```diff
@@ -53,11 +54,11 @@ public IISDeploymentParameters GetBaseDeploymentParameters(TestApplication appli
                 publish);
         }
 
-        public IISDeploymentParameters GetBaseDeploymentParameters(TestApplication application, DeploymentParameters baseParameters, bool publish = false)
+        public IISDeploymentParameters GetBaseDeploymentParameters(ApplicationPublisher application, DeploymentParameters baseParameters, bool publish = false)
```
Comment (Tratcher) [inline test/Common.FunctionalTests/PublishedSitesFixture.cs]: application -> publisher

File: test/Common.FunctionalTests/PublishedSitesFixture.cs [hunk 4]
```diff
@@ -19,18 +19,19 @@ public class PublishedSitesCollection : ICollectionFixture<PublishedSitesFixture
 
     public class PublishedSitesFixture : IDisposable
     {
-        public TestApplication InProcessTestSite { get; } = new TestApplication(Helpers.GetInProcessTestSitesPath());
-        public TestApplication OutOfProcessTestSite { get; } = new TestApplication(Helpers.GetOutOfProcessTestSitesPath());
-        public TestApplication StartupExceptionWebsite { get; } = new TestApplication(Helpers.GetTestWebSitePath("StartupExceptionWebsite"));
-        public TestApplication OverriddenServerWebSite { get; } = new TestApplication(Helpers.GetTestWebSitePath("OverriddenServerWebSite"));
-        
+        public CachingApplicationPublisher InProcessTestSite { get; } = new CachingApplicationPublisher(Helpers.GetInProcessTestSitesPath());
```
Comment (Tratcher) [inline test/Common.FunctionalTests/PublishedSitesFixture.cs]: The field type can be the base ApplicationPublisher right?

Comment (pakrym) [reply test/Common.FunctionalTests/PublishedSitesFixture.cs]: No, ApplicationPublisher doesn't have dispose.

**Comment by pakrym** [comment]
VSTS failure: https://github.com/aspnet/IISIntegration/issues/1130

**Pull request merged**